From c2565330cd5b683578bab46c22aad029a6e9f139 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Mon, 09 May 2022 18:09:18 +0300
Subject: [PATCH] [SW-85036] rom: fix efuse read address setting

Efuse read address is currenlty configured only on cold reset as the
efuse re-initialization is done based on cold reset flops.

Due to this, in other resets(except cold) the value read by rom for
mmEFUSE_READ_ADDRESS is not the soconline default value as boot manager
updates the same, which effectively shortens the efuse read ability
from firmware.

This patch fixes this by decoupling mmEFUSE_READ_ADDRESS configuration
from cold reset flops flag.

Change-Id: I238b2540953e74971888ded31b3c8c731f4bfd45
---

diff --git a/common/efuse.c b/common/efuse.c
index 03d0571..49475a9 100644
--- a/common/efuse.c
+++ b/common/efuse.c
@@ -29,6 +29,15 @@
 	int i;
 	uint32_t read_reg;
 
+	/*
+	 * Read address range is modified by boot manager, which reduces
+	 * the read range of efuse data in firmware. Hence this read address
+	 * configuration in firmware needs to be done in each reset cycle
+	 * and should not be coupled with efuse re-initialization which is
+	 * done only on cold reset.
+	 */
+	regmap_writel(EFUSE_READ_RANGE_ALL, mmEFUSE_READ_ADDRESS);
+
 	if (arcboot_rom_global.efuse_read == 1)
 		return -1;
 
@@ -36,9 +45,6 @@
 
 	/* TODO: ask suyash to test time of this function */
 	/* TODO: facto 10 of function time ~10mS*/
-	regmap_writel(EFUSE_READ_RANGE_ALL,
-		      mmEFUSE_READ_ADDRESS);
-
 	regmap_writel(1, read_reg);
 
 	for (i = 0; i < FUSE_RETRIES; i++)
